File: output/Llama-2-100/avatar/Java/Python/atcoder_AGC028_B.py
def atcoder_AGC028_B(n):
    br = input()
    sa = br.split()
    a = [int(x) for x in sa]
    mod = 1000000007
    p = 1
    q = [1] * n
    for i in range(2, n):
        p = p * (i + 1) % mod
    for i in range(1, n - 1):
        q[i] = (q[i - 1] + modinv(i + 1, mod)) % mod
    ans = 0
    for i in range(n):
        val = q[i] + q[n - i - 1] - 1
        val *= a[i]
        val %= mod
        val *= p
        val %= mod
        ans += val
    print(ans)

def modinv(a, m):
    b = m
    u = 1
    v = 0
    while b > 0:
        t = a // b
        a -= t * b
        tmp = a
        a = b
        b = tmp
        u -= t * v
        tmp = u
        u = v
        v = tmp
    u %= m
    if u < 0:
        u += m
    return u

File: output/Llama-2-100/avatar/Python/Java/atcoder_AGC028_B.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.math.BigInteger;

public class atcoder_AGC028_B {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int M = 100000;
        List<Long> A = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            String str = sc.nextLine();
            String[] arr = str.split(" ");
            for (int j = 0; j < arr.length; j++) {
                A.add(Long.parseLong(arr[j]));
            }
        }
        int Nsq = 1000;
        BigInteger[][] fact = new BigInteger[Nsq][Nsq];
        BigInteger[] inv = new BigInteger[N];
        for (int n = 1; n <= Nsq; n++) {
            fact[n][0] = BigInteger.ONE;
            for (int i = 1; i < Nsq; i++) {
                fact[n][i] = fact[n - 1][i - 1] * fact[n - 1][i];
                fact[n][i] = fact[n][i].mod(M);
            }
        }
        for (int n = 1; n <= Nsq; n++) {
            inv[n] = BigInteger.valueOf(pow(n + 1, M - 2, M));
        }
        BigInteger temp = inv[j] + inv[N - 1 - j] - BigInteger.ONE;
        BigInteger ans = (BigInteger) A.stream().map(a -> a * temp.mod(M)).sum().mod(M);
        ans = ans.multiply(fact[N]).mod(M);
        System.out.println(ans);
    }
}


